### PR TITLE
Refactoring into submodules

### PR DESCRIPTION
**Pull request by blegat**
The MOI wrapper wasn't yet aligned with the fact that we now have bridges in top of the DiffOpt models.

Closed https://github.com/jump-dev/DiffOpt.jl/issues/220

### CODE REVIEW

**Comment by matbesancon** [comment]
Why introduce submodules? Overall I find it to quickly complicate both internal and external development, I know this is common in MOI but DiffOpt is small enough to contain everything

**Comment by blegat** [comment]
We have this structure where for QP and CP, we have a `Form`, `Model`, `Cache`, ...
We need to give them different names to avoid clashes so we are adding references to `Quadratic` and `Conic` to differentiate them but it's a bit messy. The modules provide a namespace which makes it a lot cleaner.
Second, the DiffOpt optimizer is modular in the sense that it could work with any `DiffOpt.AbstractModel` subtypes that show how to differentiate for a particular model. We advertise it in the paper and the fact we can refactor cleanly with modules is a consequence of it.

Could you elaborate on why it makes things complicated ? I agree that it makes it complicated if these things are not independent but in this case, they are quite independent. We want to have a clean interface that someone can implement to add new `DiffOpt.AbstractModel`. Having these module makes this interface apparent and forces us to keep a clean one.

**Comment by matbesancon** [comment]
Major hassle with submodules is the `import ..Foo or import .Bar` but I see the point that it's not too relevant here

**Comment by blegat** [comment]
Yes, since the inclusion in the modules shorten the names, the name of the modules + name of the object does not increase too much in length

File: src/DiffOpt.jl [hunk 1]
```diff
@@ -19,47 +18,22 @@ const MOSD = MathOptSetDistances
...
-include("conic_diff.jl")
-include("quadratic_diff.jl")
 include("moi_wrapper.jl")
 include("jump_moi_overloads.jl")
 
 include("bridges.jl")
 
+include("QuadraticProgram/QuadraticProgram.jl")
+include("ConicProgram/ConicProgram.jl")
+function add_all_model_constructors(model)
```
Comment (matbesancon) [inline src/DiffOpt.jl:31]: docstring for this function?

File: src/moi_wrapper.jl [hunk 2]
```diff
@@ -165,41 +132,18 @@ function MOI.supports(model::Optimizer, attr::MOI.AbstractModelAttribute)
...
-function MOI.supports_constraint(
-    model::Optimizer,
-    ::Type{F},
-    ::Type{S}
-) where {
-    F <: SUPPORTED_VECTOR_FUNCTIONS,
-    S <: SUPPORTED_VECTOR_SETS,
-}
+    model::Optimizer, ::Type{F}, ::Type{S},
+) where {F<:MOI.AbstractFunction,S<:MOI.AbstractSet}
```
Comment (joaquimg) [inline src/moi_wrapper.jl:141]: we might need to check both optimizer and the underlying diff_models

**Review by joaquimg — APPROVED** [review]

**Comment by matbesancon** [comment]
@blegat can you check the two comments above, then good to go

**Comment by matbesancon** [comment]
bump

File: src/moi_wrapper.jl (hunk 2 above)
Comment (matbesancon) [reply src/moi_wrapper.jl:141]: ping @blegat ?

File: src/DiffOpt.jl (hunk 1 above)
Comment (blegat) [reply src/DiffOpt.jl:38]: Done

File: src/moi_wrapper.jl (hunk 2 above)
Comment (blegat) [reply src/moi_wrapper.jl:148]: I'd rather add special cases for the attributes that are used with the `DiffOpt.AbstractModel`, that's what I just did in the last commit

**Comment by matbesancon** [comment]
@blegat good to go I think? We should tag a breaking release right after

**Pull request merged**